tracks/bash/exercises/hamming/mentoring.md

```diff
@@ -0,0 +1,127 @@
+### Reasonable Solutions
+
+```bash
+#!/usr/bin/env bash
+main () {
+  (( "$#" == 2 )) || {
+      echo "Usage: hamming.sh <string1> <string2>"
+      exit 1
+  }
+  
+  # Regular vars are easier to read when doing fancy parameter expansion.
+  a=$1 b=$2 
+  
+  (( "${#a}" == "${#b}" )) || {
+      echo "left and right strands must be of equal length"
+      exit 1
+  }
+
+  declare -i count
+  for (( i = 0; i < "${#a}"; i++ )); do
+    [[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+  done
+
+  printf '%d\n' "$count"
+}
+
+main "$@"
+```
+```bash
+#!/usr/bin/env bash
+main () {
+    if (( $# != 2 )); then 
+        echo "Usage: hamming.sh <string1> <string2>"
+        exit 1
+    fi
+    if (( ${#1} != ${#2} )); then
+        echo "left and right strands must be of equal length"
+        exit 1
+    fi
+    hamming=0
+    for (( i=0; i<${#1}; i++ )); do
+        [[ "${1:$i:1}" != "${2:$i:1}" ]] && (( ++hamming ))
+    done
+    echo "$hamming"
+}
+
+main "$@"
+```
+
+
+### Common Suggestions
+
+* Suggest using a for loop in bash or other [looping constructs](https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs) https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs
+
+* Suggest using [Parameter Expansion - Substring Extraction](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) to access the index of the string. https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+
+* Suggest creating a counter to gather hammingDistance
+1) let "hammingDistance+=1"
+2) counter=0; (( ++counter ));
+3) counter=0; counter=(( $counter+1 ));
+
+* Suggest Quote the right-hand side of == in [[ ]] to prevent glob matching, the left hand side can be quoted too.
+```
+run bash hamming.sh 'AAA' 'A?A'
+
+✗ expose subtle '[[ $x == $y ]]' bug
+   (in test file hamming_test.sh, line 92)
+     `(( status == 0 ))' failed
+```
+
+Problem line of code: 
+```bash
+[[ ${a:i:1} == ${b:i:1} ]] || count+=1
+```
+Potential fix:
+```bash
+[[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+```
+
+
+### Talking Points
+
+* `if`/`else` used instead of List Constructs.
+
+Because conditionals are needed. If the
+student uses `if`/`else` it's an opportunity to talk about the list constructs for a more concise answer. `&&`/`||` The and list and or list constructs provide a means of processing a number of commands consecutively. These can effectively replace complex nested if/then or even case statements. [List Constructs](https://tldp.org/LDP/abs/html/list-cons.html#LISTCONSREF)
+
+### General Guidelines
+
+
+* there are unquoted variables
+
+```md
+On LN you have an unquoted variable! That can lead to unexpected behavior.
```

Comment:
I have no idea what "LN" means.  If it is unclear to me, it may be unclear to other mentors.  (Or it shows that I simply am not in the "in crowd" to know the jargon.)

Reply:
I think it is a placeholder for "line number". Such that the sentence "renders" into "On line 20 you have an unquoted variable!".

Reply:
Happy to change to "line number" but other mentor docs use "LN"
so unsure what to do here ¯\_(ツ)_/¯

Reply:
My feeling on abbreviations is that they should be defined locally on first use in the location.  Hopefully other mentor documents also define on first use.  Or link to a listing of abbreviations used in the context of Exercism so we don't have to guess about communication.

Reply:
@kotp , here is where I grabbed that "LN" snippet , https://github.com/exercism/website-copy/blob/master/tracks/bash/exercises/two-fer/mentoring.md#common-suggestions

There is also no definition for it in that location. 👍

Reply:
That is unfortunate that there is no definition there as well.  ;) "The benefit of copy and paste is that errors are less likely to be introduced, the drawback to copy and paste is that the same mistake is easy to duplicate."